Buff Thermic, Frost and Webslinger spiders, tweak cocoons

MuckerMayhem opened the pull request:
:cl: Mucker
rscadd: Add new abilities to the Frost and Webslinger spider variants.
tweak: Buffed the Thermic spider's poison.
bugfix: Fixed hostile mobs being able to target mobs in cocoons in some cases.
/:cl:

Full changes:

moves the cocoon logic to its Initialize() proc so it can be more easily utilized elsewhere.
sets notarget on living mobs inside cocoons.

Frost spiders:
Adds a special attack to the frost spider that lets them rapidly cool down the area they are in.
Venom now lowers body temperature by a greator amount.

Thermic:
Injects more poison per bite.

Webslinger:
The projectile now adds stacking 'webs' to the victim that slow it down and eventually cocoon it.

Just need to figure out why removing webs doesn't update the aura's icon state, and the best way to let people remove webs from themselves is, then this will be good to go. Otherwise, everything is final.

SierraKomodo commented on code/modules/mob/living/simple_animal/hostile/giant_spider/webslinger.dm:
```diff
@@ -42,3 +42,128 @@
+	if (!auras)
+		return .
+
+	for (var/obj/aura/A in auras)
+		if (istype(A, /obj/aura/web))
+			var/obj/aura/web/W = A
+			var/tally = W.stacks * 2
+
+			return . + tally
```
Inherent type checks in for loops. Less redundancy.

```suggestion
	for (var/obj/aura/web/W in auras)
		var/tally = W.stacks * 2

		return . + tally
```

SierraKomodo commented on code/modules/mob/living/simple_animal/hostile/giant_spider/webslinger.dm:
```diff
@@ -42,3 +42,128 @@
+			return . + tally
+
+/mob/living/attackby(obj/item/I, mob/user)
+	for (var/obj/aura/A in auras)
+		if (istype(A, /obj/aura/web))
+			var/obj/aura/web/W = A
+
+			if (!W.hitby(I, user))
+				return ..()
```
```suggestion
	for (var/obj/aura/web/W in auras)
		if (!W.hitby(I, user))
			return ..()
```

SierraKomodo commented on code/modules/mob/living/simple_animal/hostile/giant_spider/webslinger.dm:
```diff
@@ -42,3 +42,128 @@
+	return ..()
+
+/mob/living/carbon/attack_hand(mob/user)
+	for (var/obj/aura/A in auras)
+		if (istype(A, /obj/aura/web))
+			var/obj/aura/web/W = A
+
+			if (!W.hitby(null, user))
+				return ..()
+			return
```
```suggestion
	for (var/obj/aura/web/W in auras)
		if (!W.hitby(null, user))
			return ..()
		return
```

SierraKomodo commented on code/modules/mob/living/simple_animal/hostile/giant_spider/webslinger.dm:
```diff
@@ -50,9 +50,19 @@
+	remove.Grant(user)
+
+/obj/aura/web/Destroy()
+	if (remove)
+		remove.Remove(user)
```
```suggestion
	if (remove)
		remove.Remove(user)
		remove = null
```

SierraKomodo commented on code/modules/mob/living/simple_animal/hostile/giant_spider/webslinger.dm:
```diff
@@ -103,67 +113,65 @@
+	remove_webbing()
 
-	var/mob/living/carbon/human/H = user
+/obj/aura/web/proc/remove_webbing(var/obj/item/I, mob/living/M)
```
```suggestion
/obj/aura/web/proc/remove_webbing(obj/item/I, mob/living/M)
```

MuckerMayhem commented:
Pretty much everything is done now, just need to figure out why icons aren't updating in the case of removing webs from a person.

SierraKomodo commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -196,6 +196,10 @@
 	if(became_older)
 		msg += "[T.He] looks a lot older than you remember.\n"
 
+	for (var/obj/aura/web/W in auras)
+		msg += SPAN_WARNING("[T.He] is covered in webs!\n")
+		break
```
Out of curiosity, does `if (/obj/aura/web in auras)` work?

MuckerMayhem replied:
Gave it a shot, but no go unfortunately.

SierraKomodo commented on code/modules/mob/living/simple_animal/hostile/giant_spider/webslinger.dm:
```diff
@@ -42,3 +42,137 @@
+
+		return
+
+	icon_state = "web_[clamp(stacks, 1, 4)]"
```
`icon_state` change operations should ideally be contained within `on_icon_update()`, and anywhere that needs to make an icon_state change should call `update_icon()`.

I'd move this specific line to an `on_icon_update()` override, then put an `update_icon()` call here.

This could also allow you to have the on_update_icon override do something like this, so the icon is more fitting to whatever max_stacks is, if you want to be fancy:

```dm
var/max_icon = 4
var/stacks_ratio = stacks / max_stacks
var/icon_ratio = round(max_icon * stacks_ratio)
icon_ratio = clamp(icon_ratio, 1, max_icon) // Failsafe for rounding hitting 0, or stacks exceeding max_stacks for any reason
icon_state = "web_[icon_ratio]"
```

SierraKomodo commented on code/modules/mob/living/simple_animal/hostile/giant_spider/webslinger.dm:
```diff
@@ -42,3 +42,137 @@
+
+/obj/aura/web/proc/remove_stack()
+	stacks = stacks - 1
+	icon_state = "web_[clamp(stacks, 1, 4)]"
```
As above

SierraKomodo commented on code/modules/mob/living/simple_animal/hostile/giant_spider/webslinger.dm:
```diff
@@ -42,3 +42,137 @@
+	var/max_stacks = 5
+	var/datum/action/remove_web/remove_webs
+
+/obj/aura/web/New(mob/living/target)
+	. = ..()
+	remove_webs = new()
+	remove_webs.web = src
+	remove_webs.Grant(user)
```
Can this use `Initialize()` instead of `New()`?

MuckerMayhem replied:
Unfortunately not, otherwise `user` doesn't exist in time. The parent call could probably replace its `New()` with `Initialize()`, but that's something I can do in a future PR, along with doing mob icon updates.

MuckerMayhem committed:
buff the thermic, frost, webslinger spiders

moves the cocoon logic to its Initialize() proc so it can be more easily utilized elsewhere.
sets notarget on living mobs inside cocoons.

Frost spiders:
Adds a special attack to the frost spider that lets them rapidly cool down the area they are in.
Venom now lowers body temperature by a greator amount.

Thermic:
Injects more poison per bite.

Webslinger:
The projectile now adds stacking 'webs' to the victim that slow it down and eventually cocoon it.
